### PR TITLE
fix(intel): resolve syscall numbers via basic-block backtracking

### PR DESCRIPTION
**Pull request by r0ny123**
## Summary

SMDA's Intel backend classifies a Linux `syscall` as process-ending only when the syscall number register (`rax`==60, i.e. `exit`) is set by the **immediately preceding** `mov rax/eax, <imm>`. If the number is set a little less directly, the old code logged a debug message and gave up — that was the standing TODO in the `except ValueError` path.

Addresses upstream issue [danielplohmann/smda#119](https://github.com/danielplohmann/smda/issues/119).

## Changes

- Add `_resolveSyscallNumber(preceding_instructions, bitness)` in `src/smda/intel/IntelDisassembler.py`: a conservative, **block-local** backtracking resolver that walks the already-analyzed block instructions in reverse to find the `mov`/`movabs` that sets the syscall register.
- Add `SYSCALL_BACKTRACK_BOUNDARY` (reuses the existing `CALL/JMP/CJMP/LOOP/RET` sets + `syscall/sysenter/int/int3/hlt`) so backtracking never crosses a control-flow edge.
- Accumulate `block_instructions` across cache look-ahead refills (the disasm cache is only a ~15-byte window), so resolution isn't limited to one window and the direct case is preserved across refills.
- Replace the inline immediate parse (and its TODO) with a call to the resolver.

## Conservatism (no false endings)

The resolver returns `None` — never a guess — when the register is written by anything it can't track:
- non-`mov` writes (`xor eax, eax`, `lea`, `pop rax`, …),
- sub-register writes (`al`/`ah`/`ax`),
- register/memory/expression sources (`mov rax, rbx`, `mov rax, [rbp-8]`),
- or when a control-flow boundary is reached first.

So it only ever reports "process ending" when an actual `mov/movabs <rax|eax>, 0x3c` precedes the `syscall` in the same block. 64-bit honors the zero-extending `mov eax, 0x3c`; immediates are parsed with base 0 (handles capstone's `0x..` hex and decimal).

## Tests

`tests/testIntelDisassembler.py`: unit tests for the direct case (64- and 32-bit), backtracking over unrelated instructions, the `movabs` and zero-extending-`eax` forms, and the unresolved paths (clobber, control-flow boundary incl. a prefixed `bnd ret`, register source, empty input).

## Validation

- `ruff check .` / `ruff format --check .` clean
- `make test` → 114 passed
- Reviewed by a Sonnet 4.6 subagent (extra-high reasoning). Findings addressed: fixed a cross-cache-refill regression by accumulating `block_instructions` (instead of slicing the look-ahead window), added `movabs` handling, and switched to base-0 immediate parsing.

Refs: #119

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **Bug Fixes**
  * Improved syscall detection with conservative backtracking for more accurate identification of program termination within basic blocks.

* **Tests**
  * Added comprehensive test coverage for syscall resolution across various instruction patterns and boundary conditions.

### CODE REVIEW

File: src/smda/intel/IntelDisassembler.py [hunk 1]
```diff
@@ -342,6 +354,54 @@ def _getDisasmWindowBuffer(self, addr):
...
+                continue
+            if mnemonic in ("mov", "movabs") and len(operands) == 2 and destination in target_regs:
+                try:
+                    # base 0 handles capstone's 0x-prefixed hex as well as decimal
+                    return int(operands[1], 0)
+                except ValueError:
+                    # source is a register, memory operand, or expression -> unresolved
+                    return None
+            # any other write to the register family cannot be tracked conservatively
+            return None
```
[the remark this answers is not in the document]
Comment (r0ny123) [reply src/smda/intel/IntelDisassembler.py:463]: Fixed in 69fc9da. Both directions are now handled:

- **False positives (implicit writers):** added `SYSCALL_IMPLICIT_RAX_WRITERS` (cpuid, rdtsc(p), rdmsr/rdpmc, lods*, cbw/cwde/cdqe, lahf, xlat(b), mul/div/idiv, in(s), cmpxchg*, the legacy BCD ops), one-operand `imul`, and `xchg` when rax/eax is *any* operand — these now stop backtracking instead of being skipped.
- **False negatives (read-only):** `cmp`/`test`/`push`/`bt` are skipped rather than treated as clobbers.

This was the important class since a false positive would truncate a real function, so I erred toward stopping. Added unit tests for both.

File: tests/testIntelDisassembler.py [hunk 2]
```diff
@@ -106,6 +106,49 @@ def test_accepts_missing_timeout_callback(self):
...
+        across_boundary = [self._ins("mov", "rax, 0x3c"), self._ins("call", "0x401000")]
+        self.assertIsNone(disassembler._resolveSyscallNumber(across_boundary, 64))
+        # prefixed boundary mnemonic ("bnd ret") is still recognized after prefix split
+        across_prefixed_boundary = [self._ins("mov", "rax, 0x3c"), self._ins("bnd ret", "")]
+        self.assertIsNone(disassembler._resolveSyscallNumber(across_prefixed_boundary, 64))
+        # value sourced from a register/memory operand is not a parseable immediate
+        from_register = [self._ins("mov", "rax, rbx")]
+        self.assertIsNone(disassembler._resolveSyscallNumber(from_register, 64))
+        # no preceding instructions at all
+        self.assertIsNone(disassembler._resolveSyscallNumber([], 64))
```
[the remark this answers is not in the document]
Comment (r0ny123) [reply tests/testIntelDisassembler.py:150]: Done in 69fc9da — added `test_syscall_number_continues_past_read_only_instructions` (cmp/test/push still resolve to 60) and `test_syscall_number_unresolved_on_implicit_rax_clobber` (cpuid/rdtsc/lodsq/cdqe/div/one-op imul and `xchg [rdi], rax` all return None; multi-operand `imul rbx, rcx, 2` to an unrelated register still resolves).

File: src/smda/intel/IntelDisassembler.py [hunk 3]
```diff
@@ -342,6 +354,54 @@ def _getDisasmWindowBuffer(self, addr):
...
+        else:
+            target_regs = ("eax",)
+            clobber_regs = ("eax", "ax", "al", "ah")
+        for instruction in reversed(preceding_instructions):
+            mnemonic = instruction[2].split(" ")[-1]
+            if mnemonic in SYSCALL_BACKTRACK_BOUNDARY or mnemonic.startswith("j"):
+                return None
+            operands = [operand.strip().lower() for operand in instruction[3].split(",") if operand.strip()]
+            if not operands:
+                continue
```
[the remark this answers is not in the document]
Comment (r0ny123) [reply src/smda/intel/IntelDisassembler.py:450]: Good catch — fixed in 69fc9da. Operand-less (and other) implicit rax/eax writers like `cpuid`, `rdtsc`, `lahf`, `xlatb`, `lods*`, `cdqe`, etc. are now in `SYSCALL_IMPLICIT_RAX_WRITERS` and stop backtracking, so `mov eax,0x3c; cpuid; syscall` no longer resolves to the stale 60. Added a regression test covering this.

**Comment by r0ny123** [comment]
Addressed the CodeRabbit nitpick from [review 4406311017](https://github.com/r0ny123/smda/pull/66#pullrequestreview-4406311017): added `xgetbv` to `SYSCALL_IMPLICIT_RAX_WRITERS` in c5e04c2. Like `cpuid`/`rdtsc`/`rdpmc`, it writes `edx:eax` with no explicit destination operand, so it now stops syscall-number backtracking instead of being skipped. Added it to the implicit-clobber regression test.
